Upgrade Receive Dialog

mankyKitty opened the pull request:
This is a WIP of upgrading the receive dialog from the 'accounts' to be able to receive money from a non-bip32 key that is not stored in the wallet.

~~It's a WIP. so be gentle :hammer_and_wrench:~~

Commit:
WIP Begin upgrading the account receive dialog

Commit:
WIP receive dialog upgrade

benkolera commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -308,6 +321,33 @@ storeKeys ks = setItemStorage localStorage StoreWallet_Keys ks
 loadKeys :: (FromJSON key, HasStorage m, MonadJSM m) => m (Maybe (Accounts key))
 loadKeys = getItemStorage localStorage StoreWallet_Keys
 
+uiAccountNameInput
```
This appears to be a UI thing in Frontend.Wallet Probs should go in Frontend.UI.*

Commit:
Prevent exceptions from bubbling out from JS call

benkolera commented on frontend/src/Frontend.hs:
```diff
@@ -53,7 +54,7 @@ frontend = Frontend
     (fileOpened, triggerOpen) <- openFileDialog
     let store = browserStorage
-        crypto = Crypto mkSignature (const genKeyPair)
+        crypto = Crypto mkSignature (const genKeyPair) deriveKeyPairFromPrivateKey
```
Don't forget the same change in Desktop bipWallet! You are then gonna have to deal with XPrv there so it'll be harder. 

It may actually be worth just using the Pact / Cryptonite stuff there given we had to change the crypto api anyway...

https://github.com/kadena-io/chainweaver/pull/264/files#diff-152ee524a8f6733eb8014b53f0ebbf73R150

benkolera commented on frontend/src/Frontend/UI/Widgets.hs:
```diff
@@ -372,6 +374,30 @@ noAutofillAttrs = Map.fromList
+  -- ^ Validation function returning `Just error message` on error.
+  -> m (InputElement EventResult (DomBuilderSpace m) t)
+  -> m (Event t a)
+validatedInput uCls check input = divClass (renderClass uCls) $ do
```
You made this and didn't end up using it?

benkolera commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -1,30 +1,254 @@
-
+{-# LANGUAGE RecursiveDo #-}
```
I'm a bit suspicious of why we can't use uiDeploymentSettings here like the signing ui does. But I trust you and we're out of time to dig deeply. Code dupe and cleanup later if possible sounds fine. :slightly_smiling_face:

mankyKitty replied:
This is a suspicion that I share. The original UX flow did not allow for the deployment process to be used. That still holds but I'm wary of _not_ using `uiDeploymentSettings` for obvious reasons. Definitely something for me to follow up on.

Commit:
Re-add verification function to crypto apiv1client

rework part of receive dialog to use it

Commit:
Remove redundant imports

Commit:
Remove some redundant imports

Commit:
Drop broken preview and empty signing tabs

Commit:
Reduce the number of steps in workflow

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -1,62 +1,445 @@
+    chain <- divClass "account-details__receive-from-chain" $ userChainIdSelect model
+
+    amount <- fst . snd <$> mkLabeledInput True "Amount"
+      (uiRealWithPrecisionInputElement maxCoinPrecision id) def
```
#278 renames this function (semantic conflict)

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -1,62 +1,445 @@
+
+  (showingAddr, (conf, tfrInfo)) <- divClass "modal__main account-details" $ do
+    rec
+      showingKadenaAddress <- holdDyn True $
```
`showingKadenaAddress <- toggle True $ onAddrClick <> onReceiClick`

mankyKitty replied:
nice, ty

Commit:
Move ui widget from Frontend.Wallet

Ensure the gaslimit and ttl are passed through

alexfmpe commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -81,10 +86,18 @@ import Frontend.KadenaAddress
+accountIsCreated = maybe AccountCreated_No (const AccountCreated_Yes) . _account_balance
+
+accountCreatedYesNo :: Account key -> a -> a -> a
+accountCreatedYesNo a yes no = case accountIsCreated a of
```
How about `accountCreatedBool :: a -> a -> Account key -> a` (and flip the yes/no order) to match `bool` ?

alexfmpe commented on frontend/src/Frontend/Wallet.hs:
```diff
@@ -81,10 +86,18 @@ import Frontend.KadenaAddress
+import Frontend.UI.Widgets
+
+accountIsCreated :: Account key -> AccountCreated
+accountIsCreated = maybe AccountCreated_No (const AccountCreated_Yes) . _account_balance
```
Any reason not to use `newtype AccountCreated = AccountCreated Bool` instead?
Then you could simply `accountIsCreated = AccountCreated . isJust . _account_balance`

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -1,62 +1,446 @@
+         <@> tagMaybe (current infos) deploy
+       )
+
+-- uiReceiveDeployLegacyTransfer
```
What's going on here?

mankyKitty replied:
Nothing to see here, move along ...

Commit:
Fix up the styling a bit, remove commented code

Commit:
Changes based on PR feedback

Commit:
Removed TODO and updated name

Commit:
Add missing file

Commit:
Refactored AccountCreated as newtype

Commit:
Allow for closing of final receive dialog via top close button

Commit:
Remove debugging from network

Commit:
Nudge legacy accordion down a bit, remove debug

benkolera commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+  deriving (Generic)
+
+-- TODO: This is not backwards compatible. Do we care?
+instance Aeson.FromJSON DesktopKey
```
We want a backwards compat non derived instance here esp because we never want to store these new keys.

benkolera commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+instance Aeson.FromJSON DesktopKey
+instance Aeson.ToJSON DesktopKey
+
+-- The pact 3.0 blog said that pact supports ethereum ECDSA and ED25519 keys. We probably need
```
We don't need this comment. Doug confirmed ED25519 keys only.

benkolera commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+    DesktopKeyBip32 xprv ->
+      pure $ Newtype.pack $ Crypto.unXSignature $ Crypto.sign (T.encodeUtf8 pass) xprv bs
+    DesktopKeyPact pkScheme pubKey encryptedSecret -> do
+      -- TODO TODO TODO : Still no PBKDF2 decryption from the secret key yet.
```
Need to tidy up the names and comments here because this thing isn't and isn't planned to be encrypted.

benkolera commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+      pure (DesktopKeyBip32 xprv, unsafePublicKey $ Crypto.xpubPublicKey $ Crypto.toXPub xprv)
+    -- This is a little bastardised now as we aren't really generating anything: we are just
+    -- encrypting the secret for serialisation.
+    GenFromPactKey pactKey -> do
```
This branch should error. We absolutely never ever want to store these keys.

tomsmalley replied:
Having to error out in these branches definitely means we shouldn't go this route

benkolera commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+
+-- The pact 3.0 blog said that pact supports ethereum ECDSA and ED25519 keys. We probably need
+-- to support both here. See https://medium.com/kadena-io/announcing-pact-3-0-4b0a8f35e6a0
+bipCrypto :: Crypto.XPrv -> Text -> Crypto DesktopKey
```
Not for now, but we should drop in an asana ticket to maybe separate the type for storage vs signing. @tomsmalley  suggested maybe doing an associated type in the crypto class and that sounds pretty good to me.

mankyKitty replied:
Ticket created

benkolera commented on frontend/src/Frontend.hs:
```diff
@@ -53,7 +53,7 @@ frontend = Frontend
     (fileOpened, triggerOpen) <- openFileDialog
     let store = browserStorage
-        crypto = Crypto mkSignature (const genKeyPair)
+        crypto = Crypto mkSignature (const genKeyPair) (\_ _ -> pure $ Left "not supported")
```
You could use Frontend.Crypto.Ed25519.deriveKeyPairFromPrivateKey here, but it probably doesn't matter since it wont get used anyway.

benkolera commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+      let someKpE = importKey pkScheme (Just $ Newtype.unpack pubKey) (unXPactSecret encryptedSecret)
+      case someKpE of
+        -- TODO: Hash is supposed to be a Base64 encoded bytestring. This likely isn't right :)
+        Right someKp -> liftIO $ Newtype.pack <$> PactCrypto.sign someKp (Pact.Hash bs)
```
I'm surprised and terrified that that actually worked. Lol. Probably remove the TODO though if it worked. :)

tomsmalley commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
           path k = dir </> FilePath.makeValid (show k)
 
-bipCrypto :: Crypto.XPrv -> Text -> Crypto Crypto.XPrv
+newtype XPactSecret = XPactSecret { unXPactSecret :: BS.ByteString }
```
what is this, is it different from `PrivateKey`?

mankyKitty replied:
It's going away, along with `DesktopKey`

tomsmalley commented on desktop/src/Desktop/Frontend.hs:
```diff
@@ -92,15 +99,62 @@ fileStorage dir = Storage
+  -- TODO: Should there be some verification of length like XPrv?
+  parseJSON = fmap (XPactSecret . fst . B16.decode . T.encodeUtf8) . Aeson.parseJSON
+
+data DesktopKey
```
The presence of this seems odd if we aren't storing things. Like you said in mattermost, signing with these keys can be done by changing `buildCmd` only.

mankyKitty replied:
Removing

tomsmalley commented on frontend/src/Frontend/Crypto/Class.hs:
```diff
@@ -15,19 +15,44 @@ import Obelisk.Route.Frontend
   { _crypto_sign :: ByteString -> key -> JSM Signature
-  , _crypto_genKey :: Int -> JSM (key, PublicKey)
+  , _crypto_genKey :: GenKeyArg -> JSM (key, PublicKey)
+  , _crypto_verifyPactKey :: PPKScheme -> Text -> JSM (Either String PactKey)
```
This seems like it doesn't belong here. Does this function change between desktop/web?

tomsmalley replied:
Yes, it does change: on desktop we use the function from Pact.Types.Crypto which is only available on GHC.

tomsmalley commented on frontend/src/Frontend/Crypto/Class.hs:
```diff
@@ -15,19 +15,44 @@ import Obelisk.Route.Frontend
 data Crypto key = Crypto
   { _crypto_sign :: ByteString -> key -> JSM Signature
-  , _crypto_genKey :: Int -> JSM (key, PublicKey)
+  , _crypto_genKey :: GenKeyArg -> JSM (key, PublicKey)
```
Shouldn't need to do this if we change `buildCmd`

mankyKitty replied:
Reverting

tomsmalley commented on frontend/src/Frontend/KadenaAddress.hs:
```diff
@@ -107,11 +109,12 @@ data KadenaAddressPrefix
   deriving (Show, Eq)
 
+accountCreatedBool :: a -> a -> AccountCreated -> a
+accountCreatedBool no yes (AccountCreated b) = bool no yes b
```
Boolean blindness here is terrible, I'd rather just pattern match anywhere this is used

tomsmalley commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -1,62 +1,337 @@
+      }
+
+    transferCapability = DappCap
+      { _dappCap_role = "TRANFER"
```
Spelling

tomsmalley commented on frontend/src/Frontend/UI/Dialogs/Receive.hs:
```diff
@@ -1,62 +1,337 @@
-  done <- modalFooter $ confirmButton def "Done"
-  pure (mempty, close <> done)
+  cmd <- buildCmd Nothing networkId pm signingPairs [] code dat pkCaps
+  pure $ ReceiveBuildCommandInfo code transferInfo cmd signingPairs pm caps pkCaps dat
```
Why package all this info up and then only use the command at the use site?

Commit:
Updates based on PR feedback

Remove desktop key type as we're never storing it so we don't need the distinction and the
potential for errors was too great.

Add build command function for building a command with a PactKey as a signer.

Revert changes to AccountCreated back to sum type.

Remove accountCreatedBool function.

Tidy up the Receive module to remove unnecessary functions.
